### DISCUSSION TITLE
A second comparison chart of SD Samplers vs Steps, using landscapes this time

### DISCUSSION THREAD

**Discussion by JoshuaKimsey** [hlky/stable-diffusion-webui#178] [Show and tell]
Wanted to share this here in case @hlky wanted to add it to their guide, or anyone else wants to use it as well. Wanted to do landscapes as it provides a nice contrast to the portrait-style comparison used in the other chart. Hopefully, it helps additionally show the strengths and weaknesses of each sampler vs the steps needed to get a good output. :)

Prompt: "The Island of Ireland at Sunset, by Albert Bierstadt, Fine Art, Good, Sunlight, Rembrandt Lighting"
Seed: 2058446450
CFG: 10

![Stable-Diffusion-Landscape-Comparison](https://user-images.githubusercontent.com/22421310/187063145-3d4f16d7-7bd6-4804-be1c-acf228ed2507.jpg)

**Comment by hlky** [comment]
What are you using to make this? a collection of these for different prompts would be useful

I think a wiki page all about prompt crafting would help a lot of people too

**Reply by JoshuaKimsey** [reply]
I just made this by hand with GIMP. Took the image you uploaded to your guide, overlaid the images I generated, and changed the CFG scale at the top. 

The prompt was made using [promptoMANIA](https://promptomania.com/), as it makes crafting a prompt for a desired style really easy.
(They also now recommend your fork of SD in their most recent change log, as well.)

Another great source for searching for prompt styles is [Lexica.art](https://lexica.art/). Which is basically a repository of prompts with the art styles they produced. These are the two main one's I'm aware of, though I'm sure more will follow.

However, I was actually a bit curious if a system to automatically generate a grid like the one above could be implemented? Obviously it would be a bit more niche, and core functionality should come first. But, having the ability to auto-generate a grid showing what a particular prompt, seed, and CFG combo would like across samplers might be a nice addition for people searching for their desired output.

**Reply by hlky** [reply]
Yeah it's possible to do it automatically, probably a little time consuming to implement though

**Reply by JoshuaKimsey** [reply]
Yeah, definitely shouldn't be a priority, but would be a nice addition later on.

**Reply by ABBEN4** [reply]
I think it is doable on colab, been playing with random generations and it leads to a nice grid. Just need to make the same but at each round of generation change the number of steps and the sampler. Might look at this if you are interested in that or for prompt engineers !

**Reply by hlky** [reply]
If you can get some code to reproduce a similar grid to @JoshuaKimsey then that would be great, if not don't worry, I'll figure it out

**Reply by bitRAKE** [reply]
How important is time in this? All steps don't appear equal in the temporal domain.

**Reply by ABBEN4** [reply]
I check that out, is it possible to call the different samplers directly from here ? : 

with autocast("cuda"):
    chosen = pipe(prompt, num_inference_steps=inf, generator=generator, height=sizeh, width=sizew, guidance_scale=Guide)["sample"] 

Not been playing with the samplers yet, but if yes, I will make an auto comparison colab to fix certain parameters and vary those we want to compare. Could be usefull to see the effect of others parameters in a systematic way too.

**Comment by micwalk** [comment]
https://github.com/404james/stable-diffusion-grids has a janky but functional CLI way to generate grids like these. Would be cool to integrate into the UI properly instead of the hack it currently is.

**Comment by dudumaroja** [comment]
i, for one, love the euler_a Style

**Reply by yourjelly** [reply]
Absolutely, however the fact that it changes dramatically with steps makes it difficult to manage for buffing out fine details.

**Comment by thedyze** [comment]
Would going above 64 steps serve any purpose?

**Reply by JoshuaKimsey** [reply]
It does, it evens out smaller details and can make things look sharper. That said, it is  a situation of diminishing returns. I typically do test generations at about 100 steps, 125 if it's something a bit more detailed, then do a full render at 250 steps for a final image.

This is in regards to the `k_` samplers, btw. `DDIM` and `PLMS` may actually need well over 100 steps to settle on a final form, compared to `k_lms` that can do so reasonably well by 50, or `k_euler` that can often do so by 30.

**Reply by thedyze** [reply]
I tend to favorite k_euler a lot. It does tend to do a bit "wavey" lines on patters where there should be straight lines, is there any sampler that does patterns/lines more cleanly than the rest?

**Reply by EliEron** [reply]
As a bit of an aside it's important to note that the _a family of samplers (k_euler_a, k_dpm_2_a) behave quite differently when it comes to step counts than the other samplers.

Rather than just vary the amount of detail they tend to produce wildly different images at different step counts. 

k_euler_a 32 steps:
![k_euler_a_32-2](https://user-images.githubusercontent.com/22575945/188340879-320a83e9-8b8b-44f7-b038-05a00cfd4a4f.png)


k_euler_a 128 steps:
![k_euler_a_128-2](https://user-images.githubusercontent.com/22575945/188340886-db2e68e4-5c34-45ac-9431-8ac567964411.png)

And here is k_euler (non _a) in contrast:

k_euler 32 steps:
![k_euler_32-2](https://user-images.githubusercontent.com/22575945/188341290-babb1c30-97c1-4093-a64d-86230366bc08.png)

k_euler 128 steps:
![k_euler_128-2](https://user-images.githubusercontent.com/22575945/188341296-dfb3092d-c76e-49dd-b733-f9f3c572fc9c.png)

As you can see k_euler_a produces entirely different images, whereas k_euler mostly just changes the details in the image it had generated. Though it's worth stating that this is not a hard rule. I have experienced major changes with non _a samplers with different steps, but it is very rare in contrast to the _a samplers where it is very common.

This comparison also brings up another important point, the _a samplers produce images that are quite different from the other samplers. Most samplers will produce images that resemble each other when ran with the same parameters (though there will be notable differences) but the _a samplers will usually produce images that look nothing like the images produced by the other samplers. Though the _a samplers are consistent with each other. So k_euler_a and k_dpm_2_a will produce somewhat similar images when provided with the same parameters.

**Reply by mchaker** [reply]
Thank you for this insight @EliEron

**Comment by protoplm** [comment]
Related feature request to have something which could streamline the analysis of what parameters do to a prompt, and some examples from what I'm trying to do with it.

https://github.com/sd-webui/stable-diffusion-webui/discussions/1242
